proxy/proxy.go

```diff
@@ -401,8 +401,20 @@ func (p *Proxy) Resolve(d *DNSContext) error {
 	// Use cache only if it's enabled and the query doesn't use custom
 	// upstreams.
 	cacheWorks := p.cache != nil && d.CustomUpstreamConfig == nil
-	if cacheWorks {
-		if p.replyFromCache(d) {
+	if cacheWorks && !d.disableCacheReading {
+		hit, isNotExpired := p.replyFromCache(d)
+		hit = hit && (p.Config.CachePrefetch || isNotExpired)
+		if hit {
```

Comment:
Perhaps it would be better to not reassign `hit` and just write:

```go
if hit && (p.Config.CachePrefetch || isNotExpired) {
        // …
}
```

Reply:
fixed